### PR TITLE
[Feature:System] Add support for ldap authentication

### LINKED ISSUE #2054: Implement LDAP Authentication
Like #1803, we should try and integrate other well-known authentication systems that schools have setup for authentication within their computer systems. LDAP is one such protocol that should be relatively straight forward to create.

### PR DESCRIPTION
**Pull request by MasterOdin**
### Please check if the PR fulfills these requirements:

* [x] The PR title and message follows our [guidelines](https://submitty.org/developer/how_to_contribute)
* [x] Tests for the changes have been added/updated (if possible)
* [x] Documentation has been updated/added if relevant: https://github.com/Submitty/submitty.github.io/pull/385

### What is the current behavior?

Closes #2054 

Submitty does not currently support using LDAP to authenticate users.

### What is the new behavior?

Adds LDAP support for Submitty.

### CODE REVIEW

File: .setup/CONFIGURE_SUBMITTY.py [hunk 1]
```diff
@@ -536,6 +562,18 @@ def write(x=''):
...
 
+##############################################################################
+# Write authentication json
+if not args.worker:
+    config = OrderedDict()
+    config['authentication_method'] = AUTHENTICATION_METHOD
+    config['ldap_options'] = LDAP_OPTIONS
+
+    with open(AUTHENTICATION_JSON, 'w') as json_file:
+        json.dump(config, json_file, indent=2)
```
[the remark this answers is not in the document]
Comment (MasterOdin) [reply .setup/CONFIGURE_SUBMITTY.py:573]: What do you mean? All changes in this file (and throughout the PR) have 4 spaces for indentation it looks like?

[the remark this answers is not in the document]
Comment (williamjallen) [reply .setup/CONFIGURE_SUBMITTY.py:573]: Perhaps I misunderstood then.  I thought the `indent=2` parameter indented each level of the JSON by 2 spaces.

[the remark this answers is not in the document]
Comment (MasterOdin) [reply .setup/CONFIGURE_SUBMITTY.py:573]: Oh, I thought you were talking about indentation in the python file, not the argument. You're right, this should be `indent=4`.

As a note, this is a great usage of the `suggestion` PR comment, like so:
````
```suggestion
        json.dump(config, json_file, indent=2)
```
````

This gets rendered as:

```suggestion
        json.dump(config, json_file, indent=4)
```

(Not sure if you knew about this)

File: migration/migrator/migrations/system/20210918211432_authentication_method.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    auth_method = authentication['authentication_method']
+    if auth_method == 'LdapAuthentication':
+        auth_method = 'PamAuthentication'
+    with (config.config_path / 'database.json').open('r+') as db_file:
+        db_info = json.load(db_file, object_pairs_hook=OrderedDict)
+        db_file.seek(0)
+        db_file.truncate()
+
+        db_info['authentication_method'] = auth_method
+        json.dump(db_info, db_file, indent=2)
```
Comment (williamjallen) [inline migration/migrator/migrations/system/20210918211432_authentication_method.py:57]:
```suggestion
        json.dump(db_info, db_file, indent=4)
```

File: migration/migrator/migrations/system/20210918211432_authentication_method.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            auth_file.seek(0)
+            auth_file.truncate()
+            auth_file.write(json.dumps(auth_info, indent=2))
+
+    with (config.config_path / 'database.json').open('r+') as db_file:
+        db_info = json.load(db_file, object_pairs_hook=OrderedDict)
+        db_file.seek(0)
+        db_file.truncate()
+        del db_info['authentication_method']
+        json.dump(db_info, db_file, indent=2)
```
Comment (williamjallen) [inline migration/migrator/migrations/system/20210918211432_authentication_method.py:32]:
```suggestion
        json.dump(db_info, db_file, indent=4)
```

File: migration/migrator/migrations/system/20210918211432_authentication_method.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    if not authentication_file.exists():
+        with authentication_file.open('w') as auth_file:
+            json.dump({'authentication_method': config.database['authentication_method'], 'ldap_options': []}, auth_file, indent=2)
+    else:
+        with authentication_file.open('r+') as auth_file:
+            auth_info = json.load(auth_file, object_pairs_hook=OrderedDict)
+            auth_info['authentication_method'] = config.database['authentication_method']
+            auth_file.seek(0)
+            auth_file.truncate()
+            auth_file.write(json.dumps(auth_info, indent=2))
```
Comment (williamjallen) [inline migration/migrator/migrations/system/20210918211432_authentication_method.py:25]:
```suggestion
            auth_file.write(json.dumps(auth_info, indent=4))
```

File: migration/migrator/migrations/system/20210918211432_authentication_method.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    :param config: Object holding configuration details about Submitty
+    :type config: migrator.config.Config
+    :param database: Object for interacting with given database for environment
+    :type database: migrator.db.Database
+    """
+    authentication_file = config.config_path / 'authentication.json'
+    if not authentication_file.exists():
+        with authentication_file.open('w') as auth_file:
+            json.dump({'authentication_method': config.database['authentication_method'], 'ldap_options': []}, auth_file, indent=2)
```
Comment (williamjallen) [inline migration/migrator/migrations/system/20210918211432_authentication_method.py:18]:
```suggestion
            json.dump({'authentication_method': config.database['authentication_method'], 'ldap_options': []}, auth_file, indent=4)
```

**Review by williamjallen — CHANGES_REQUESTED** [review]

**Comment by MasterOdin** [comment]
Running a final vagrant up test here: https://github.com/Submitty/Submitty/actions/runs/1492144889

**Pull request merged**